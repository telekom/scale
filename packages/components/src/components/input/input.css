/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

.input {
  position: relative;
}
.input .input__helper-text,
.input .input__counter {
  font-weight: var(--scl-font-weight-bold);
}
.input .input__input,
.input .input__select {
  width: 100%;
  height: var(--scl-spacing-48);
  margin: 0;
  display: flex;
  outline: none;
  padding: var(--scl-spacing-12) var(--scl-spacing-12);
  z-index: 1;
  box-sizing: border-box;
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  font-family: inherit;
  font-size: var(--scl-font-size-16);
  border-radius: var(--scl-radius-8);
  border: var(--scl-spacing-1) solid var(--scl-color-black);
}
.input .input__textarea {
  width: 100%;
  margin: 0;
  resize: vertical;
  display: flex;
  outline: none;
  padding: var(--scl-spacing-12) var(--scl-spacing-12);
  z-index: 1;
  box-sizing: border-box;
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  font-family: inherit;
  font-size: var(--scl-font-size-16);
  border-radius: var(--scl-radius-8);
  border: var(--scl-spacing-1) solid var(--scl-color-text-standard);
}
.input .input__select {
  appearance: none;
  -webkit-appearance: none;
}
.input .input__select-wrapper {
  position: relative;
}
.input .input__counter {
  display: flex;
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  margin-left: auto;
  padding-right: var(--scl-spacing-12);
  justify-content: flex-end;
  font-size: var(--scl-font-size-12);
  line-height: var(--scl-font-line-height-133);
  color: var(--scl-color-text-standard);
}
.input .input__helper-text {
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  padding-left: var(--scl-spacing-12);
  font-size: var(--scl-font-size-12);
  line-height: var(--scl-font-line-height-133);
  color: var(--scl-color-blue-70);
}
.input .input__meta {
  display: flex;
  justify-content: space-between;
  margin-top: var(--scl-spacing-4);
}
.input:not(.input--disabled):hover .input__select-wrapper {
  --icon-color: var(--scl-color-primary-hover, #f90984);
}
.input:not(.input--disabled):active .input__select-wrapper {
  --icon-color: var(--scl-color-primary-active, #cb0068);
}
.input:not(.input--disabled) .input__input:hover,
.input:not(.input--disabled) .input__input:focus,
.input:not(.input--disabled) .input__select:hover,
.input:not(.input--disabled) .input__select:focus,
.input:not(.input--disabled) .input__textarea:hover,
.input:not(.input--disabled) .input__textarea:focus {
  border-color: var(--scl-color-primary-hover, #f90984);
}
.input:not(.input--disabled) .input__input:focus,
.input:not(.input--disabled) .input__select:focus,
.input:not(.input--disabled) .input__textarea:focus {
  box-shadow: 0 0 0 var(--scl-spacing-2) var(--scl-color-focus);
}
.input:not(.input--disabled) .input__input:focus::placeholder,
.input:not(.input--disabled) .input__select:focus::placeholder,
.input:not(.input--disabled) .input__textarea:focus::placeholder {
  color: var(--scl-color-grey-60);
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
}
.input .input__select-wrapper scale-icon {
  top: 50%;
  right: var(--scl-spacing-12);
  position: absolute;
  transform: translateY(-50%);
  pointer-events: none;
}
.input .input__input::placeholder,
.input .input__select::placeholder,
.input .input__textarea::placeholder,
.input ::placeholder {
  color: transparent;
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
}
.input--variant-static .input__label {
  color: var(--scl-color-grey-60);
  display: flex;
  font-weight: var(--scl-font-weight-medium);
}
.input--variant-animated .input__input,
.input--variant-animated .input__select {
  padding: var(--scl-spacing-12) var(--scl-spacing-12) 0
    calc(var(--scl-spacing-12) - 1px);
}
.input--variant-animated .input__textarea {
  padding-top: var(--scl-spacing-24);
}
.input--variant-animated .input__label {
  top: 0;
  left: 0;
  color: var(--scl-color-grey-60);
  display: flex;
  z-index: var(--scl-z-index-10);
  position: absolute;
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  pointer-events: none;
  font-size: var(--scl-font-size-16);
  /* previously: transform: translate(13px, calc((48px - 1rem) / 2)); */
  transform: translate(
    var(--scl-spacing-12),
    calc((var(--scl-spacing-48) - var(--scl-font-size-16)) / 2)
  );
  font-weight: var(--scl-font-weight-medium);
}
.input--variant-animated.input--has-focus .input__label,
.input--variant-animated.animated .input__label {
  color: var(--scl-color-grey-60);
  transform: translate(var(--scl-spacing-12), var(--scl-spacing-8));
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  font-size: var(--scl-font-size-10);
  font-weight: var(--scl-font-weight-bold);
}
.input--status-error .input__input,
.input--status-error .input__textarea,
.input--status-error .input__select {
  border: var(--scl-spacing-2) solid var(--scl-color-background-error);
}
.input--status-error .input__helper-text {
  color: var(--scl-color-text-error);
}
.input--status-error .input__counter {
  color: var(--scl-color-text-error);
}
.input--size-small .input__input {
  height: var(--scl-spacing-40);
}
.input--size-small .input__select {
  height: var(--scl-spacing-40);
}
.input--size-small .input__label {
  top: 0;
  left: 0;
  color: var(--scl-color-grey-60);
  display: flex;
  z-index: var(--scl-z-index-10);
  position: absolute;
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  pointer-events: none;
  font-size: var(--scl-font-size-16);
  /* previously: transform: translate(
    13px,
    calc((var(--input-small-height, 40px) - 1rem) / 2)
  ); */
  transform: translate(
    var(--scl-spacing-12),
    calc((var(--scl-spacing-40) - var(--scl-font-size-16)) / 2)
  );
  font-weight: var(--scl-font-weight-medium);
}
.input--size-small.input--has-focus .input__label,
.input--size-small.animated .input__label {
  color: var(--scl-color-grey-60);
  transform: translate(var(--scl-spacing-12), var(--scl-spacing-4));
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  font-size: var(--scl-font-size-10);
  font-weight: var(--scl-font-weight-medium);
}
.input--transparent .input__input,
.input--transparent .input__textarea,
.input--transparent.input--type-radio .input__radio,
.input--transparent .input__select {
  background-color: transparent;
}
.input--type-checkbox {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
}
.input--type-checkbox input {
  width: 0;
  height: 0;
  opacity: 0;
  position: absolute;
}
.input--type-checkbox .input__meta {
  width: 100%;
}
.input--type-checkbox .input__helper-text {
  padding-left: var(--scl-spacing-32);
}
.input--type-checkbox label {
  color: var(--scl-color-text-standard);
  font-weight: var(--scl-font-weight-medium);
}
.input--type-checkbox .input__checkbox-container {
  width: var(--scl-spacing-16);
  height: var(--scl-spacing-16);
  display: flex;
  position: relative;
  align-items: center;
  margin-right: var(--scl-spacing-8);
}
.input--type-checkbox input:disabled ~ label {
  /* FIXME make sure color is right */
  color: var(--scl-color-background-disabled);
}
.input--type-checkbox input:checked:disabled ~ label {
  color: var(--scl-color-background-disabled);
}
.input--type-checkbox
  input:checked:disabled
  ~ .input__checkbox-container
  .input__checkbox-placeholder {
  background: var(--scl-color-background-disabled);
}
/* TODO this could removed once scale-icon is removed */
.input--type-checkbox
  input:checked:disabled
  ~ .input__checkbox-container
  scale-icon {
  --icon-color: var(--scl-color-grey-30);
}
.input--type-checkbox
  input:checked:not([disabled]):hover
  ~ .input__checkbox-container
  .input__checkbox-placeholder,
.input--type-checkbox
  input:checked:not([disabled])
  ~ .input__checkbox-container:hover
  .input__checkbox-placeholder {
  box-shadow: none;
  border-color: var(--scl-color-primary-hover, #f90984);
  background: var(--scl-color-primary-hover, #f90984);
}
.input--type-checkbox
  input:checked:not([disabled]):active
  ~ .input__checkbox-container
  .input__checkbox-placeholder,
.input--type-checkbox
  input:checked:not([disabled])
  ~ .input__checkbox-container:active
  .input__checkbox-placeholder {
  border-color: var(--scl-color-primary-active, #cb0068);
  background: var(--scl-color-primary-active, #cb0068);
}
/* TODO this could removed once scale-icon is removed */
.input--type-checkbox
  input:checked:not([disabled]):active
  ~ .input__checkbox-container
  scale-icon,
.input--type-checkbox
  input:checked:not([disabled])
  ~ .input__checkbox-container:active
  scale-icon {
  --icon-color: var(--scl-color-white);
}
.input--type-checkbox
  input:checked:not([disabled])
  ~ .input__checkbox-container
  .input__checkbox-placeholder {
  border: var(--scl-spacing-1) solid var(--scl-color-primary, #e20074);
  background: var(--scl-color-primary, #e20074);
}
.input--type-checkbox
  input:disabled
  ~ .input__checkbox-container
  .input__checkbox-placeholder {
  border-color: var(--scl-color-background-disabled);
}
.input--type-checkbox
  input:focus
  ~ .input__checkbox-container
  .input__checkbox-placeholder {
  box-shadow: 0 0 0 var(--scl-spacing-2) var(--scl-color-focus);
}
.input--type-checkbox
  input:not([disabled]):hover
  ~ .input__checkbox-container
  .input__checkbox-placeholder,
.input--type-checkbox
  input:not([disabled])
  ~ .input__checkbox-container:hover
  .input__checkbox-placeholder {
  box-shadow: none;
  border-color: var(--scl-color-primary-hover, #f90984);
}
.input--type-checkbox
  input:not([disabled]):hover
  ~ .input__checkbox-container
  ~ label,
.input--type-checkbox
  input:not([disabled])
  ~ .input__checkbox-container:hover
  ~ label {
  color: var(--scl-color-primary-hover, #f90984);
}
.input--type-checkbox
  input:not([disabled]):active
  ~ .input__checkbox-container
  .input__checkbox-placeholder,
.input--type-checkbox
  input:not([disabled])
  ~ .input__checkbox-container:active
  .input__checkbox-placeholder {
  border-color: var(--scl-color-primary-active, #cb0068);
  background: var(--scl-color-primary-active, #cb0068);
}
/* TODO this could removed once scale-icon is removed */
.input--type-checkbox
  input:not([disabled]):active
  ~ .input__checkbox-container
  scale-icon,
.input--type-checkbox
  input:not([disabled])
  ~ .input__checkbox-container:active
  scale-icon {
  --icon-color: var(--scl-color-primary-active, #cb0068);
}
.input--type-checkbox
  input:not([disabled]):active
  ~ .input__checkbox-container
  ~ label,
.input--type-checkbox
  input:not([disabled])
  ~ .input__checkbox-container:active
  ~ label {
  color: var(--scl-color-primary-active, #cb0068);
}
.input--type-checkbox .input__checkbox-container .input__checkbox-placeholder {
  width: var(--scl-spacing-16);
  height: var(--scl-spacing-16);
  margin: 0;
  box-sizing: border-box;
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  border-radius: var(--scl-radius-8);
  border: var(--scl-spacing-1) solid var(--scl-color-text-standard);
  background: var(--scl-color-background-standard);
}
.input--type-checkbox .input__checkbox-container scale-icon {
  top: calc(0.5 * var(--scl-spacing-12));
  left: calc(0.5 * var(--scl-spacing-12));
  width: var(--scl-spacing-12);
  height: var(--scl-spacing-12);
  position: absolute;
  user-select: none;
  /* TODO this could removed once scale-icon is removed */
  --icon-color: var(--scl-color-background-standard);
}
.input--type-checkbox .input__checkbox-container ~ label {
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
}
.input--type-checkbox.input--status-error
  .input__checkbox-container
  .input__checkbox-placeholder {
  border: var(--scl-spacing-2) solid var(--scl-color-background-error);
}
.input--type-radio {
  display: flex;
  flex-wrap: wrap;
  align-items: center;
}
.input--type-radio .input__meta {
  width: 100%;
}
.input--type-radio .input__helper-text {
  margin-top: var(--scl-spacing-4);
  padding-left: var(--scl-spacing-24);
}
.input--type-radio label {
  color: var(--scl-color-text-standard);
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  font-weight: var(--scl-font-weight-medium);
}
.input--type-radio input {
  width: var(--scl-spacing-16);
  height: var(--scl-spacing-16);
  transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  border-radius: 50%;
  -webkit-appearance: none;
  background-color: #fff;
  border: var(--scl-spacing-1) solid var(--scl-color-text-standard);
  margin: 0 var(--scl-spacing-8) 0 0;
}
.input--type-radio input:focus {
  outline: none;
  box-shadow: 0 0 0 var(--scl-spacing-2) var(--scl-color-focus);
}
.input--type-radio:hover input:not(:checked):not([disabled]) {
  box-shadow: none;
  border-color: var(--scl-color-primary-hover, #f90984);
}
.input--type-radio:hover input:not(:checked):not([disabled]) ~ label {
  color: var(--scl-color-primary-hover, #f90984);
}
.input--type-radio input:active {
  border: var(--scl-spacing-8) solid var(--scl-color-primary-active);
}
.input--type-radio input:not(:checked):not([disabled]):active ~ label {
  color: var(--scl-color-primary-active);
}
.input--type-radio input:disabled {
  border: var(--scl-spacing-1) solid var(--scl-color-background-disabled);
}
.input--type-radio input:disabled ~ label {
  color: var(--scl-color-grey-30);
}
.input--type-radio.input--status-error input {
  border: var(--scl-spacing-2) solid var(--scl-color-background-error);
}
.input--type-radio input:checked {
  border: calc(0.5 * var(--scl-spacing-12)) solid var(--scl-color-primary);
}
.input--type-radio input:checked:active {
  border: calc(0.5 * var(--scl-spacing-12)) solid
    var(--scl-color-primary-active, #cb0068);
}
.input--type-radio input:checked:disabled {
  background: var(--scl-color-grey-30);
  border: calc(0.5 * var(--scl-spacing-12)) solid
    var(--scl-color-background-disabled);
}
.input--type-radio input:checked:disabled ~ label {
  color: var(--scl-color-grey-30);
}
.input__textarea-label-safety-background {
  top: var(--scl-spacing-2);
  left: var(--scl-spacing-2);
  right: var(--scl-spacing-2);
  position: absolute;
  pointer-events: none;
  border-radius: var(--scl-radius-8);
  height: var(--scl-spacing-24);
  background-color: var(--scl-color-background-standard, #ffffff);
}
.input--disabled .input__textarea-label-safety-background {
  background-color: transparent;
}
.input--disabled label,
.input--disabled .input__label,
.input--disabled input,
.input--disabled .input__input,
.input--disabled .input__checkbox-container,
.input--disabled .input__radio,
.input--disabled .input__select,
.input--disabled .input__textarea,
.input--disabled .input__helper-text {
  cursor: not-allowed;
  border-color: var(--scl-color-background-disabled);
  color: var(--scl-color-background-disabled);
  background: var(--scl-color-white);
}
/* TODO this could removed once scale-icon is removed */
.input--disabled .input__select-wrapper {
  --icon-color: var(--scl-color-background-disabled);
}
.input--disabled.animated label.input__label {
  color: var(--scl-color-background-disabled);
}
