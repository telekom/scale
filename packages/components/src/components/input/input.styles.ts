import { JssStyle } from 'jss';

export const styles: JssStyle = {
  input: {
    '& input': {
      display: 'flex',
      width: '100%',
      borderRadius: '4px',
      border: '1px solid #cecece',
      padding: '0.5rem',
      fontSize: '16px',
      fontFamily: 'inherit',
      boxSizing: 'border-box',
      '&:not(.tabbing):focus': {
        outline: 0,
      },
    },
    '& .input__counter': {
      display: 'flex',
      justifyContent: 'flex-end',
      fontSize: '12px',
    },
    '& .input__helper-text': {
      fontSize: '9px',
    },
  },
  'input--status-error': {
    '& input': {
      border: '1px solid red',
    },
    '& .input__helper-text': {
      color: 'red',
    },
  },
  'input--variant-inline': {
    position: 'relative',
    '& label': {
      color: '#dfdfdf',
      height: '12px',
      position: 'absolute',
      top: 0,
      left: 0,
      transform: 'translate(0.5rem, 16px) scale(1)',
      zIndex: 1,
      transition: 'all 0.3s ease-in-out',
    },
    '& input': {
      marginTop: '24px',
      height: '48px',
    },
    '& input:focus + label': {
      transform: 'translate(0.5rem, 0) scale(0.75)',
      color: '#000',
      transition: 'all 0.3s ease-in-out',
    },
  },
  'input--variant-inline-animated': {
    '& label': {
      transform: 'translate(0.5rem, 0) scale(0.75)',
      color: '#000',
      transition: 'all 0.3s ease-in-out',
    },
  },
  'input--disabled': {
    '& label, & input': {
      opacity: '0.5',
      cursor: 'not-allowed',
    },
  },
};
