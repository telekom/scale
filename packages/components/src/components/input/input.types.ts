export type InputTypes =
  | 'email'
  | 'hidden'
  | 'number'
  | 'password'
  | 'tel'
  | 'text'
  | 'url';
