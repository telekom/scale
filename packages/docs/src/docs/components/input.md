---
title: Input
description: Input element
url: /docs/input
contributors:
  - eeegor
  - steffilankes
  - SylviaGonschior
---

# Input

Input component

selector: `t-input`

## Default

```javascript
<t-input></t-input>
```

<div class="demo-container">
  <t-input></t-input>
</div>

## With default value

```javascript
<t-input value="Hello"></t-input>
```

<div class="demo-container">
  <t-input value="Hello"></t-input>
</div>

## Types

```javascript
<t-input type="email"></t-input>
```

<div class="demo-container">
  <t-input type="email"></t-input>
</div>

## Number

```javascript
<t-input type="number"></t-input>
```

<div class="demo-container">
  <t-input type="number"></t-input>
</div>

## Password

```javascript
<t-input type="password"></t-input>
```

<div class="demo-container">
  <t-input type="password"></t-input>
</div>

## Telephone number

```javascript
<t-input type="tel"></t-input>
```

<div class="demo-container">
  <t-input type="tel"></t-input>
</div>

## Url

```javascript
<t-input type="url"></t-input>
```

<div class="demo-container">
  <t-input type="url"></t-input>
</div>

## Hidden

```javascript
<t-input type="hidden"></t-input>
```

<div class="demo-container">
  <t-input type="hidden"></t-input>
</div>
