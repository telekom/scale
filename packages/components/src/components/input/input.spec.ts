/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage } from '@stencil/core/testing';
import { Input } from './input';

describe('Input', () => {
  describe('snapshots', () => {
    it('should match default snapshot (type="text")', async () => {
      const page = await newSpecPage({
        components: [Input],
        html: `<scale-input></scale-input>`,
      });
      expect(page.root).toMatchSnapshot();
    });

    it('should match snapshot with slot', async () => {
      const page = await newSpecPage({
        components: [Input],
        html: `<scale-input icon="icon"><div>test</div></scale-input>`,
      });
      page.root.type = 'select';
      await page.waitForChanges();
      expect(page.root).toMatchSnapshot();
    });

    it('should match snapshot with type="checkbox"', async () => {
      const page = await newSpecPage({
        components: [Input],
        html: `<scale-input type="checkbox"></scale-input>`,
      });
      await page.waitForChanges();
      expect(page.root).toMatchSnapshot();
    });

    it('should match snapshot with type="radio"', async () => {
      const page = await newSpecPage({
        components: [Input],
        html: `<scale-input type="radio"></scale-input>`,
      });
      await page.waitForChanges();
      expect(page.root).toMatchSnapshot();
    });

    it('should match snapshot with type="textarea"', async () => {
      const page = await newSpecPage({
        components: [Input],
        html: `<scale-input type="textarea"></scale-input>`,
      });
      await page.waitForChanges();
      expect(page.root).toMatchSnapshot();
    });

    it('should match snapshot with type="input"', async () => {
      const page = await newSpecPage({
        components: [Input],
        html: `<scale-input type="input"></scale-input>`,
      });
      await page.waitForChanges();
      expect(page.root).toMatchSnapshot();
    });

    it('should match snapshot with type="email"', async () => {
      const page = await newSpecPage({
        components: [Input],
        html: `<scale-input type="email"></scale-input>`,
      });
      await page.waitForChanges();
      expect(page.root).toMatchSnapshot();
    });

    it('should match snapshot with type="hidden"', async () => {
      const page = await newSpecPage({
        components: [Input],
        html: `<scale-input type="hidden"></scale-input>`,
      });
      await page.waitForChanges();
      expect(page.root).toMatchSnapshot();
    });

    it('should match snapshot with type="number"', async () => {
      const page = await newSpecPage({
        components: [Input],
        html: `<scale-input type="number"></scale-input>`,
      });
      await page.waitForChanges();
      expect(page.root).toMatchSnapshot();
    });

    it('should match snapshot with type="password"', async () => {
      const page = await newSpecPage({
        components: [Input],
        html: `<scale-input type="password"></scale-input>`,
      });
      await page.waitForChanges();
      expect(page.root).toMatchSnapshot();
    });

    it('should match snapshot with type="tel"', async () => {
      const page = await newSpecPage({
        components: [Input],
        html: `<scale-input type="tel"></scale-input>`,
      });
      await page.waitForChanges();
      expect(page.root).toMatchSnapshot();
    });

    it('should match snapshot with type="url"', async () => {
      const page = await newSpecPage({
        components: [Input],
        html: `<scale-input type="url"></scale-input>`,
      });
      await page.waitForChanges();
      expect(page.root).toMatchSnapshot();
    });

    it('should handle inputId with value null', async () => {
      const page = await newSpecPage({
        components: [Input],
        html: `
          <scale-input 
           input-id
           >
          </scale-input>`,
      });
      expect(page.rootInstance.inputId).toBe('');
    });
  });

  describe('props', () => {
    it('should reflect attributes', async () => {
      const page = await newSpecPage({
        components: [Input],
        html: `<scale-input></scale-input>`,
      });
      page.root.type = 'select';
      page.root.name = 'name';
      page.root.variant = 'animated';
      page.root.label = 'label';
      page.root.size = 'small';
      page.root.helperText = 'helper text';
      page.root.invalid = true;
      page.root.disabled = false;
      page.root.required = true;
      page.root.checked = true;
      page.root.resize = 'none';
      page.root.value = '';
      page.root.multiple = false;
      page.root.transparent = false;
      page.root.controlled = false;
      await page.waitForChanges();
      expect(page.rootInstance.type).toBe('select');
      expect(page.rootInstance.name).toBe('name');
      expect(page.rootInstance.variant).toBe('animated');
      expect(page.rootInstance.label).toBe('label');
      expect(page.rootInstance.size).toBe('small');
      expect(page.rootInstance.helperText).toBe('helper text');
      expect(page.rootInstance.invalid).toBe(true);
      expect(page.rootInstance.disabled).toBe(false);
      expect(page.rootInstance.required).toBe(true);
      expect(page.rootInstance.checked).toBe(true);
      expect(page.rootInstance.resize).toBe('none');
      expect(page.rootInstance.value).toBe('');
      expect(page.rootInstance.multiple).toBe(false);
      expect(page.rootInstance.transparent).toBe(false);
      expect(page.rootInstance.controlled).toBe(false);
      expect(page.root).toMatchSnapshot();
    });
  });

  describe('css', () => {
    it('should handle css classes', () => {
      const element = new Input();
      element.type = 'number';
      element.hasFocus = true;
      element.checked = true;
      element.disabled = true;
      element.transparent = true;
      element.invalid = true;
      element.size = 'small';
      element.variant = 'static';
      element.value = 'animated';
      expect(element.getCssClassMap()).toContain('input--type-number');
      expect(element.getCssClassMap()).toContain('input--has-focus');
      expect(element.getCssClassMap()).toContain('input--checked');
      expect(element.getCssClassMap()).toContain('input--disabled');
      expect(element.getCssClassMap()).toContain('input--variant-static');
      expect(element.getCssClassMap()).toContain('input--transparent');
      expect(element.getCssClassMap()).toContain('input--status-error');
      expect(element.getCssClassMap()).toContain('input--variant-static');
      expect(element.getCssClassMap()).toContain('animated');
    });
  });

  describe('functions', () => {
    it('inputId is set on componentWillLoad', () => {
      const element = new Input();
      element.componentWillLoad();
      expect(element.inputId.substr(0, 6)).toBe('input-');
    });

    it('icon is default icon on componentWillLoad', () => {
      const element = new Input();
      element.type = 'select';
      element.icon = null;
      element.componentWillLoad();
      expect(element.icon).toBe(
        'M20.65 7.4c-.3-.3-.75-.3-1.05 0L12 15 4.4 7.4c-.3-.3-.75-.3-1.05 0s-.3.75 0 1.05L12 17.1l8.65-8.65c.3-.25.3-.75 0-1.05z'
      );
    });

    it('test function handleCheckboxClick()', () => {
      const element = new Input();
      element.disabled = false;
      element.checked = true;
      element.handleCheckboxClick();
      expect(element.checked).toBe(false);
      element.disabled = true;
      element.checked = true;
      element.handleCheckboxClick();
      expect(element.checked).toBe(true);
    });

    it('test function emitChange()', () => {
      const element = new Input();
      element.emitChange();
      expect(element.scaleChange.emit).toBeTruthy();
    });
  });

  describe('events', async () => {
    let page;
    beforeEach(async () => {
      page = await newSpecPage({
        components: [Input],
        html: `<scale-input></scale-input>`,
      });
    });

    it('onFocus type="select"', async () => {
      const mock = jest.fn();
      const mockLegacy = jest.fn();
      page.root.type = 'select';
      page.root.addEventListener('scale-focus', mock);
      page.root.addEventListener('scaleFocus', mockLegacy);
      await page.waitForChanges();
      const select = page.root.querySelector('.input__select');
      await page.waitForChanges();
      select.dispatchEvent(new Event('focus'));
      await page.waitForChanges();
      expect(mock).toHaveBeenCalled();
      expect(mockLegacy).toHaveBeenCalled();
    });

    it('onBlur type="select"', async () => {
      const mock = jest.fn();
      const mockLegacy = jest.fn();
      page.root.type = 'select';
      page.root.addEventListener('scale-blur', mock);
      page.root.addEventListener('scaleBlur', mockLegacy);
      await page.waitForChanges();
      const select = page.root.querySelector('.input__select');
      await page.waitForChanges();
      select.dispatchEvent(new Event('blur'));
      await page.waitForChanges();
      expect(mock).toHaveBeenCalled();
      expect(mockLegacy).toHaveBeenCalled();
    });

    it('onKeyDown type="select"', async () => {
      const mock = jest.fn();
      const mockLegacy = jest.fn();
      page.root.type = 'select';
      page.root.addEventListener('scale-keydown', mock);
      page.root.addEventListener('scaleKeyDown', mockLegacy);
      await page.waitForChanges();
      const select = page.root.querySelector('.input__select');
      await page.waitForChanges();
      select.dispatchEvent(new Event('keydown'));
      await page.waitForChanges();
      expect(mock).toHaveBeenCalled();
      expect(mockLegacy).toHaveBeenCalled();
    });

    it('onChange type="checkbox"', async () => {
      page.root.type = 'checkbox';
      page.root.checked = 'true';
      await page.waitForChanges();
      const select = await page.root.querySelector(
        'div.input--type-checkbox > input'
      );
      await select.dispatchEvent(new Event('change'));
      await page.waitForChanges();
      expect(page.rootInstance.checked).toBe(true);
    });

    it('onInput type="textarea"', async () => {
      page.root.type = 'textarea';
      page.root.value = 'value';
      const mock = jest.fn();
      const mockLegacy = jest.fn();
      page.root.addEventListener('scale-change', mock);
      page.root.addEventListener('scaleChange', mockLegacy);
      await page.waitForChanges();
      const select = await page.root.querySelector('textarea.input__textarea');
      await select.dispatchEvent(new Event('input'));
      await page.waitForChanges();
      expect(page.rootInstance.value).toBe('');
      expect(mock).toHaveBeenCalled();
      expect(mockLegacy).toHaveBeenCalled();
    });

    it('onChange type="textarea"', async () => {
      page.root.type = 'textarea';
      page.root.value = 'value';
      const mock = jest.fn();
      const mockLegacy = jest.fn();
      page.root.addEventListener('scale-change', mock);
      page.root.addEventListener('scaleChange', mockLegacy);
      await page.waitForChanges();
      const select = await page.root.querySelector('textarea.input__textarea');
      await select.dispatchEvent(new Event('change'));
      await page.waitForChanges();
      expect(page.rootInstance.value).toBe('');
      expect(mock).toHaveBeenCalled();
      expect(mockLegacy).toHaveBeenCalled();
    });
  });
});
