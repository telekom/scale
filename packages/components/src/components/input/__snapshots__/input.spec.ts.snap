// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should match snapshot 1`] = `
<scale-input>
  <mock:shadow-root>
    <style>
      .input-0-0-7 {
  position: relative;
}
.input-0-0-7 .input__input {
  width: 100%;
  border: 1px solid #cecece;
  height: 48px;
  display: flex;
  padding: 12px 12px;
  z-index: 1;
  font-size: 16px;
  box-sizing: border-box;
  transition: all 0.2s ease-in-out;
  font-family: inherit;
  border-radius: 4px;
}
.input-0-0-7 .input__counter {
  color: #333;
  display: flex;
  font-size: 12px;
  transition: all 0.2s ease-in-out;
  margin-left: auto;
  justify-content: flex-end;
}
.input-0-0-7 .input__helper-text {
  font-size: 12px;
  transition: all 0.2s ease-in-out;
}
.input-0-0-7 .input__meta {
  display: flex;
  margin-top: 6px;
  justify-content: space-between;
}
.input-0-0-7:not(.input--disabled-0-0-12) {}
.input-0-0-7:not(.input--disabled-0-0-12) .input__input {}
.input-0-0-7:not(.input--disabled-0-0-12) .input__input:hover, .input-0-0-7:not(.input--disabled-0-0-12) .input__input:focus {
  border-color: #2962ff;
}
.input--variant-static-0-0-8 {}
.input--variant-static-0-0-8 .input__label {
  display: flex;
  margin-bottom: 12px;
}
.input--variant-animated-0-0-9 {}
.input--variant-animated-0-0-9 .input__input.has-label {
  padding: 12px 12px 0 11px;
}
.input--variant-animated-0-0-9 .input__label {
  top: 0;
  left: 0;
  color: #dfdfdf;
  display: flex;
  z-index: 10;
  position: absolute;
  font-size: 16px;
  transform: translate(13px, 16px);
  transition: all 0.2s ease-in-out;
  font-weight: 400;
  pointer-events: none;
}
.input--variant-animated-0-0-9 .input__input:focus + .input__label, .input--variant-animated-0-0-9.animated .input__label {
  font-size: 10px;
  transform: translate(12px, 8px);
  transition: all 0.2s ease-in-out;
  font-weight: 700;
  color: #2962ff;
}
.input--status-error-0-0-10 {}
.input--status-error-0-0-10.animated .input__label {
  color: #f50057 !important;
}
.input--status-error-0-0-10 .input__input:focus + .input__label {
  color: #f50057 !important;
}
.input--status-error-0-0-10 .input__input {
  border: 1px solid #f50057;
}
.input--status-error-0-0-10 .input__helper-text {
  color: #f50057;
}
.input--status-error-0-0-10 .input__counter {
  color: #f50057;
}
.input--size-small-0-0-11 {}
.input--size-small-0-0-11 .input__input {
  height: 40px;
}
.input--size-small-0-0-11 .input__label {
  top: 0;
  left: 0;
  color: #dfdfdf;
  display: flex;
  z-index: 10;
  position: absolute;
  font-size: 16px;
  transform: translate(13px, 12px);
  transition: all 0.2s ease-in-out;
  font-weight: 400;
  pointer-events: none;
}
.input--size-small-0-0-11 .input__input:focus + .input__label, .input--size-small-0-0-11.animated .input__label {
  font-size: 10px;
  transform: translate(12px, 4px);
  transition: all 0.2s ease-in-out;
  font-weight: 700;
  color: #2962ff;
}
.input--disabled-0-0-12 {}
.input--disabled-0-0-12 .input__label, .input--disabled-0-0-12 .input__input {
  cursor: not-allowed;
  opacity: 0.5;
}
    </style>
    <div class="input--variant-animated-0-0-9 input-0-0-7">
      <input class="input__input" placeholder="" type="text">
    </div>
  </mock:shadow-root>
</scale-input>
`;
